--- original
+++ modified
@@ -135,7 +135,7 @@
 
 // Allocate a new opaque flag context.
 //
-// Setting `program_name` is needed so `flag_c_parse(c, argc, argv)` does not try to consume
+// Setting `program_name` is needed so `flag_c_parse(c, argc, argv)` does not try to consume
 // extra argument from `argv` as it usually does. If that's exactly what you want just pass
 // `NULL` as the `program_name` then.
 //
@@ -183,7 +183,8 @@
 
 //////////////////////////////
 
-#ifdef FLAG_IMPLEMENTATION
+#if defined(FLAG_IMPL) && !defined(FLAG_IMPL_C)
+#define FLAG_IMPL_C
 
 typedef enum {
     FLAG_BOOL = 0,
@@ -968,7 +969,7 @@
     flag_c_print_error(&flag_global_context, stream);
 }
 
-#endif // FLAG_IMPLEMENTATION
+#endif // FLAG_IMPL
 
 /*
    Revision history:
@@ -1012,4 +1013,4 @@
 // AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 // LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
-// IN THE SOFTWARE.
+// IN THE SOFTWARE.
\ No newline at end of file
