--- vendor/patches/nob.h	2026-09-01 05:10:03
+++ vendor/nob.h	2026-09-01 05:09:22
@@ -363,11 +363,36 @@
             while ((expected_capacity) > (da)->capacity) {                                 \
                 (da)->capacity *= 2;                                                       \
             }                                                                              \
+            (da)->items = NOB_DECLTYPE_CAST((da)->items)NOB_REALLOC((da)->items, (da)->capacity * sizeof(*(da)->items)); \
+            NOB_ASSERT((da)->items != NULL && "Buy more RAM lol");                         \
+        }                                                                                  \
+    } while (0)
+
+    // Reserve specific size to a dynamic array
+#define nob_da_realloc_capacity(da, target_capacity)                                       \
+    do {                                                                                   \
+        if ((target_capacity) > (da)->capacity) {                                          \
+            (da)->capacity = (target_capacity);                                            \
             (da)->items = NOB_DECLTYPE_CAST((da)->items)NOB_REALLOC((da)->items, (da)->capacity * sizeof(*(da)->items)); \
             NOB_ASSERT((da)->items != NULL && "Buy more RAM lol");                         \
         }                                                                                  \
     } while (0)
 
+// Trim array's capacity to it's count
+#define nob_da_realloc_trim(da)                                                       \
+    do {                                                                              \
+      if ((da)->capacity != (da)->count) {                                            \
+        if ((da)->count == 0) {                                                       \
+          NOB_FREE((da)->items);                                                      \
+          (da)->items = NULL;                                                         \
+          (da)->capacity = 0;                                                         \
+        } else {                                                                      \
+          (da)->items = NOB_REALLOC((da)->items, (da)->count * sizeof(*(da)->items)); \
+          (da)->capacity = (da)->count;                                               \
+        }                                                                             \
+      }                                                                               \
+    } while (0)
+
 // Append an item to a dynamic array
 #define nob_da_append(da, item)                \
     do {                                       \
@@ -438,6 +463,7 @@
 #define nob_swap(T, a, b) do { T t = a; a = b; b = t; } while (0)
 
 NOBDEF bool nob_read_entire_file(const char *path, Nob_String_Builder *sb);
+NOBDEF bool nob_read_entire_stdin(Nob_String_Builder *sb);
 NOBDEF int nob_sb_appendf(Nob_String_Builder *sb, const char *fmt, ...) NOB_PRINTF_FORMAT(2, 3);
 // Pads the String_Builder (sb) to the desired word size boundary with 0s.
 // Imagine we have sb that contains 5 `a`-s:
@@ -974,7 +1000,7 @@
 //     printf(SV_Fmt" => %zu\n", SV_Arg(c), n);
 // }
 // ```
-static const uint8_t nob_bytes_for_utf8[] = {
+static const uint8_t nob__bytes_for_utf8[] = {
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
@@ -984,6 +1010,7 @@
     2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2, 2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
     3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3, 4,4,4,4,4,4,4,4,5,5,5,5,6,6,6,6,
 };
+#define nob_bytes_for_utf8(sv) nob__bytes_for_utf8[(uint8_t)*(sv).data]
 
 NOBDEF size_t nob_sv_utf8_len(Nob_String_View sv, size_t *bytes_overrun);
 
@@ -995,7 +1022,8 @@
 
 #endif // NOB_H_
 
-#ifdef NOB_IMPLEMENTATION
+#if defined(NOB_IMPL) && !defined(NOB_IMPL_C)
+#define NOB_IMPL_C
 
 // This is like nob_proc_wait() but waits asynchronously. Depending on the platform ms means different thing.
 // On Windows it means timeout. On POSIX it means for how long to sleep after checking if the process exited,
@@ -1024,7 +1052,7 @@
             if (bytes_overrun) *bytes_overrun = i - sv.count;
             return n;
         }
-        i += nob_bytes_for_utf8[(uint8_t)sv.data[i]];
+        i += nob__bytes_for_utf8[(uint8_t)sv.data[i]];
         n += 1;
     }
     NOB_UNREACHABLE("sv_utf8_len");
@@ -2597,7 +2625,28 @@
     if (f) fclose(f);
     return result;
 }
+
+NOBDEF bool nob_read_entire_stdin(Nob_String_Builder *sb)
+{
+    bool result = true;
+#ifdef _WIN32
+    if (_setmode(_fileno(stdin), _O_BINARY) < 0) nob_return_defer(false);
+#endif
 
+    char buffer[32 * 1024];
+    size_t bytes_read = 0;
+    while ((bytes_read = fread(buffer, 1, sizeof(buffer), stdin)) > 0) {
+        nob_da_append_many(sb, buffer, bytes_read);
+    }
+    if (ferror(stdin)) {
+        nob_return_defer(false);
+    }
+
+defer:
+    if (!result) nob_log(NOB_ERROR, "Could not read stdin: %s", strerror(errno));
+    return result;
+}
+
 NOBDEF int nob_sb_appendf(Nob_String_Builder *sb, const char *fmt, ...)
 {
     va_list args;
@@ -3016,6 +3065,8 @@
         #define da_append_many nob_da_append_many
         #define da_resize nob_da_resize
         #define da_reserve nob_da_reserve
+        #define da_realloc_capacity nob_da_realloc_capacity
+        #define da_realloc_trim nob_da_realloc_trim
         #define da_last nob_da_last
         #define da_first nob_da_first
         #define da_pop nob_da_pop
@@ -3028,6 +3079,7 @@
         #endif // __cplusplus
         #define String_Builder Nob_String_Builder
         #define read_entire_file nob_read_entire_file
+        #define read_entire_stdin nob_read_entire_stdin
         #define sb_appendf nob_sb_appendf
         #define sb_append_buf nob_sb_append_buf
         #define sb_append_sv nob_sb_append_sv
