--- original
+++ modified
@@ -63,7 +63,7 @@
       from other languages (for whatever reason).
 
       If only few specific names create conflicts for you, you can just #undef those names after the
-      `#include <nob.h>` without enabling `NOB_UNSTRIP_PREFIX` since they are macros anyway.
+      `#include <nob.h>` without enabling `NOB_UNSTRIP_PREFIX` since they are macros anyway.
 
    # Macro Interface
 
@@ -203,12 +203,12 @@
 #    define NOB_PRINTF_FORMAT(STRING_INDEX, FIRST_TO_CHECK)
 #endif
 
-NOBDEF void nob__panicf(const char *file, int line, const char *label, const char *format, ...);
+NOBDEF void nob__panicf(const char *file, int line, const char *label, const char *format, ...) NOB_PRINTF_FORMAT(4, 5);
 
 #define NOB_UNUSED(value) (void)(value)
-#define NOB_TODO(message) do { fprintf(stderr, "%s:%d: TODO: %s\n", __FILE__, __LINE__, message); abort(); } while(0)
+#define NOB_TODO(message) nob__panicf(__FILE__, __LINE__, "TODO", "%s", message)
 #define NOB_TODOF(...) nob__panicf(__FILE__, __LINE__, "TODO", __VA_ARGS__)
-#define NOB_UNREACHABLE(message) do { fprintf(stderr, "%s:%d: UNREACHABLE: %s\n", __FILE__, __LINE__, message); abort(); } while(0)
+#define NOB_UNREACHABLE(message) nob__panicf(__FILE__, __LINE__, "UNREACHABLE", "%s", message)
 #define NOB_UNREACHABLEF(...) nob__panicf(__FILE__, __LINE__, "UNREACHABLE", __VA_ARGS__)
 
 #define NOB_ARRAY_LEN(array) (sizeof(array)/sizeof(array[0]))
@@ -438,6 +438,7 @@
 #define nob_swap(T, a, b) do { T t = a; a = b; b = t; } while (0)
 
 NOBDEF bool nob_read_entire_file(const char *path, Nob_String_Builder *sb);
+NOBDEF bool nob_read_entire_stdin(Nob_String_Builder *sb);
 NOBDEF int nob_sb_appendf(Nob_String_Builder *sb, const char *fmt, ...) NOB_PRINTF_FORMAT(2, 3);
 // Pads the String_Builder (sb) to the desired word size boundary with 0s.
 // Imagine we have sb that contains 5 `a`-s:
@@ -503,6 +504,9 @@
     size_t count;
     size_t capacity;
 } Nob_Procs;
+
+// Wait until the process has finished and write down exit code
+NOBDEF bool nob_proc_wait_with_exit_status(Nob_Proc proc, unsigned int *exit_code);
 
 // Wait until the process has finished
 NOBDEF bool nob_proc_wait(Nob_Proc proc);
@@ -542,6 +546,7 @@
     const char *stdout_path;
     // Redirect stderr to file
     const char *stderr_path;
+    unsigned int *exit_status;
 } Nob_Cmd_Opt;
 
 // Run the command with options.
@@ -831,40 +836,6 @@
 #ifndef nob_cc_inputs
 #  define nob_cc_inputs(cmd, ...) nob_cmd_append(cmd, __VA_ARGS__)
 #endif // nob_cc_inputs
-
-// TODO: add MinGW support for Go Rebuild Urself™ Technology and all the nob_cc_* macros above
-//   Musializer contributors came up with a pretty interesting idea of an optional prefix macro which could be useful for
-//   MinGW support:
-//   https://github.com/tsoding/musializer/blob/b7578cc76b9ecb573d239acc9ccf5a04d3aba2c9/src_build/nob_win64_mingw.c#L3-L9
-// TODO: Maybe instead NOB_REBUILD_URSELF macro, the Go Rebuild Urself™ Technology should use the
-//   user defined nob_cc_* macros instead?
-#ifndef NOB_REBUILD_URSELF
-#  if defined(_WIN32)
-#    if defined(__clang__)
-#      if defined(__cplusplus)
-#        define NOB_REBUILD_URSELF(binary_path, source_path) "clang", "-x", "c++", "-o", binary_path, source_path
-#      else
-#        define NOB_REBUILD_URSELF(binary_path, source_path) "clang", "-x", "c", "-o", binary_path, source_path
-#      endif
-#    elif defined(__GNUC__)
-#      if defined(__cplusplus)
-#        define NOB_REBUILD_URSELF(binary_path, source_path) "gcc", "-x", "c++", "-o", binary_path, source_path
-#      else
-#        define NOB_REBUILD_URSELF(binary_path, source_path) "gcc", "-x", "c", "-o", binary_path, source_path
-#      endif
-#    elif defined(_MSC_VER)
-#       define NOB_REBUILD_URSELF(binary_path, source_path) "cl.exe", nob_temp_sprintf("/Fe:%s", (binary_path)), source_path
-#    elif defined(__TINYC__)
-#       define NOB_REBUILD_URSELF(binary_path, source_path) "tcc", "-o", binary_path, source_path
-#    endif
-#  else
-#    if defined(__cplusplus)
-#      define NOB_REBUILD_URSELF(binary_path, source_path) "cc", "-x", "c++", "-o", binary_path, source_path
-#    else
-#      define NOB_REBUILD_URSELF(binary_path, source_path) "cc", "-x", "c", "-o", binary_path, source_path
-#    endif
-#  endif
-#endif
 
 // Go Rebuild Urself™ Technology
 //
@@ -974,7 +945,7 @@
 //     printf(SV_Fmt" => %zu\n", SV_Arg(c), n);
 // }
 // ```
-static const uint8_t nob_bytes_for_utf8[] = {
+static const uint8_t nob__bytes_for_utf8[] = {
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
@@ -984,6 +955,7 @@
     2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2, 2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
     3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3, 4,4,4,4,4,4,4,4,5,5,5,5,6,6,6,6,
 };
+#define nob_bytes_for_utf8(sv) nob__bytes_for_utf8[(uint8_t)*(sv).data]
 
 NOBDEF size_t nob_sv_utf8_len(Nob_String_View sv, size_t *bytes_overrun);
 
@@ -995,7 +967,8 @@
 
 #endif // NOB_H_
 
-#ifdef NOB_IMPLEMENTATION
+#if defined(NOB_IMPL) && !defined(NOB_IMPL_C)
+#define NOB_IMPL_C
 
 // This is like nob_proc_wait() but waits asynchronously. Depending on the platform ms means different thing.
 // On Windows it means timeout. On POSIX it means for how long to sleep after checking if the process exited,
@@ -1024,7 +997,7 @@
             if (bytes_overrun) *bytes_overrun = i - sv.count;
             return n;
         }
-        i += nob_bytes_for_utf8[(uint8_t)sv.data[i]];
+        i += nob__bytes_for_utf8[(uint8_t)sv.data[i]];
         n += 1;
     }
     NOB_UNREACHABLE("sv_utf8_len");
@@ -1117,7 +1090,17 @@
     const char *old_binary_path = nob_temp_sprintf("%s.old", binary_path);
 
     if (!nob_rename(binary_path, old_binary_path)) exit(1);
+#ifdef NOB_REBUILD_URSELF
     nob_cmd_append(&cmd, NOB_REBUILD_URSELF(binary_path, source_path));
+#else
+    nob_cc(&cmd);
+    nob_cc_flags(&cmd);
+    nob_cc_inputs(&cmd, source_path);
+    nob_cc_output(&cmd, binary_path);
+#ifdef NOB_REBUILD_URSELF_FLAGS
+    nob_cmd_append(&cmd, NOB_REBUILD_URSELF_FLAGS);
+#endif
+#endif
     Nob_Cmd_Opt opt = {0};
     if (!nob_cmd_run_opt(&cmd, opt)) {
         nob_rename(old_binary_path, binary_path);
@@ -1358,7 +1341,7 @@
         if (proc == NOB_INVALID_PROC) nob_return_defer(false);
         nob_da_append(opt.async, proc);
     } else {
-        if (!nob_proc_wait(proc)) nob_return_defer(false);
+        if (!nob_proc_wait_with_exit_status(proc, opt.exit_status)) nob_return_defer(false);
     }
 
 defer:
@@ -1785,7 +1768,7 @@
     return nob_procs_flush(procs);
 }
 
-NOBDEF bool nob_proc_wait(Nob_Proc proc)
+NOBDEF bool nob_proc_wait_with_exit_status(Nob_Proc proc, unsigned int *exit_status_storage)
 {
     if (proc == NOB_INVALID_PROC) return false;
 
@@ -1801,6 +1784,9 @@
     }
 
     DWORD exit_status;
+    if (exit_status_storage != NULL) {
+        *exit_status_storage = exit_status;
+    }
     if (!GetExitCodeProcess(proc, &exit_status)) {
         nob_log(NOB_ERROR, "could not get process exit code: %s", nob_win32_error_message(GetLastError()));
         return false;
@@ -1824,6 +1810,9 @@
 
         if (WIFEXITED(wstatus)) {
             int exit_status = WEXITSTATUS(wstatus);
+            if (exit_status_storage != NULL) {
+                *exit_status_storage = exit_status;
+            }
             if (exit_status != 0) {
                 nob_log(NOB_ERROR, "command exited with exit code %d", exit_status);
                 return false;
@@ -1842,6 +1831,11 @@
 #endif
 }
 
+NOBDEF bool nob_proc_wait(Nob_Proc proc)
+{
+    return nob_proc_wait_with_exit_status(proc, NULL);
+}
+
 static int nob__proc_wait_async(Nob_Proc proc, int ms)
 {
     if (proc == NOB_INVALID_PROC) return false;
@@ -2597,7 +2591,28 @@
     if (f) fclose(f);
     return result;
 }
+
+NOBDEF bool nob_read_entire_stdin(Nob_String_Builder *sb)
+{
+    bool result = true;
+#ifdef _WIN32
+    if (_setmode(_fileno(stdin), _O_BINARY) < 0) nob_return_defer(false);
+#endif
 
+    char buffer[32 * 1024];
+    size_t bytes_read = 0;
+    while ((bytes_read = fread(buffer, 1, sizeof(buffer), stdin)) > 0) {
+        nob_da_append_many(sb, buffer, bytes_read);
+    }
+    if (ferror(stdin)) {
+        nob_return_defer(false);
+    }
+
+defer:
+    if (!result) nob_log(NOB_ERROR, "Could not read stdin: %s", strerror(errno));
+    return result;
+}
+
 NOBDEF int nob_sb_appendf(Nob_String_Builder *sb, const char *fmt, ...)
 {
     va_list args;
@@ -3028,6 +3043,7 @@
         #endif // __cplusplus
         #define String_Builder Nob_String_Builder
         #define read_entire_file nob_read_entire_file
+        #define read_entire_stdin nob_read_entire_stdin
         #define sb_appendf nob_sb_appendf
         #define sb_append_buf nob_sb_append_buf
         #define sb_append_sv nob_sb_append_sv
@@ -3395,4 +3411,4 @@
    ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
    WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
    ------------------------------------------------------------------------------
-*/
+*/
\ No newline at end of file
