--- original/nob.h	2026-09-01 02:40:00
+++ ./nob.h	2026-09-01 02:37:03
@@ -366,6 +366,31 @@
             (da)->items = NOB_DECLTYPE_CAST((da)->items)NOB_REALLOC((da)->items, (da)->capacity * sizeof(*(da)->items)); \
             NOB_ASSERT((da)->items != NULL && "Buy more RAM lol");                         \
         }                                                                                  \
+    } while (0)
+
+    // Reserve specific size to a dynamic array
+#define nob_da_realloc_capacity(da, target_capacity)                                       \
+    do {                                                                                   \
+        if ((target_capacity) > (da)->capacity) {                                          \
+            (da)->capacity = (target_capacity);                                            \
+            (da)->items = NOB_DECLTYPE_CAST((da)->items)NOB_REALLOC((da)->items, (da)->capacity * sizeof(*(da)->items)); \
+            NOB_ASSERT((da)->items != NULL && "Buy more RAM lol");                         \
+        }                                                                                  \
+    } while (0)
+
+// Trim array's capacity to it's count
+#define nob_da_realloc_trim(da)                                                       \
+    do {                                                                              \
+      if ((da)->capacity != (da)->count) {                                            \
+        if ((da)->count == 0) {                                                       \
+          NOB_FREE((da)->items);                                                      \
+          (da)->items = NULL;                                                         \
+          (da)->capacity = 0;                                                         \
+        } else {                                                                      \
+          (da)->items = NOB_REALLOC((da)->items, (da)->count * sizeof(*(da)->items)); \
+          (da)->capacity = (da)->count;                                               \
+        }                                                                             \
+      }                                                                               \
     } while (0)
 
 // Append an item to a dynamic array
@@ -995,7 +1020,8 @@
 
 #endif // NOB_H_
 
-#ifdef NOB_IMPLEMENTATION
+#if defined(NOB_IMPL) && !defined(NOB_IMPL_C)
+#define NOB_IMPL_C
 
 // This is like nob_proc_wait() but waits asynchronously. Depending on the platform ms means different thing.
 // On Windows it means timeout. On POSIX it means for how long to sleep after checking if the process exited,
@@ -3016,6 +3042,8 @@
         #define da_append_many nob_da_append_many
         #define da_resize nob_da_resize
         #define da_reserve nob_da_reserve
+        #define da_realloc_capacity nob_da_realloc_capacity
+        #define da_realloc_trim nob_da_realloc_trim
         #define da_last nob_da_last
         #define da_first nob_da_first
         #define da_pop nob_da_pop
